src/compiler/checker.ts

```diff
@@ -888,11 +888,17 @@ namespace ts {
             return result;
         }
 
-        function mergeSymbol(target: Symbol, source: Symbol) {
+        /**
+         * Note: if target is transient, then it is mutable, and mergeSymbol with both mutate and return it.
+         * If target is not transient, mergeSymbol will produce a transient clone, mutate that and return it.
+         */
+        function mergeSymbol(target: Symbol, source: Symbol): Symbol {
             if (!(target.flags & getExcludedSymbolFlags(source.flags)) ||
                 (source.flags | target.flags) & SymbolFlags.JSContainer) {
-                const targetValueDeclaration = target.valueDeclaration;
-                Debug.assert(!!(target.flags & SymbolFlags.Transient));
+                if (!(target.flags & SymbolFlags.Transient)) {
+                    target = cloneSymbol(target);
+                }
+                Debug.assert(source !== target);

```

Comment:
Should this be above the `Transitent` check, or is it OK for a nontransient target to merge into "itself" (its transient copy)?

Reply:
Strictly speaking, it doesn't matter, because source will never equal target just after running cloneSymbol. I think it is clearer beforehand though.